relay-server/src/actors/outcome.rs

```diff
@@ -622,31 +704,144 @@ impl Actor for HttpOutcomeProducer {
+        self.send_http_message(msg, ctx);
+        Ok(())
+    }
+}
 
-impl Default for HttpOutcomeProducer {
+impl Handler<TrackOutcome> for HttpOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, ctx: &mut Self::Context) -> Self::Result {
+        self.handle(TrackRawOutcome::from_outcome(msg, &self.config), ctx)
+    }
+}
+
+struct ClientReportOutcomeProducer {}
+
+impl Actor for ClientReportOutcomeProducer {
+    type Context = Context<Self>;
+}
+
+impl Handler<TrackOutcome> for ClientReportOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, _ctx: &mut Self::Context) -> Self::Result {
+        let mut client_report = ClientReport {
+            timestamp: Some(UnixTimestamp::from_secs(
+                msg.timestamp.timestamp().try_into().unwrap_or(0),
+            )),
+            ..Default::default()
+        };
+
+        // The outcome type determines what field to place the outcome in:
+        let discarded_events = match msg.outcome {
+            Outcome::Filtered(_) => &mut client_report._server_filtered_events,
+            Outcome::FilteredSampling(_) => &mut client_report._server_filtered_sampling_events,
+            Outcome::RateLimited(_) => &mut client_report._server_rate_limited_events,
+            _ => {
+                // Cannot convert this outcome to a client report.
+                return Ok(());
+            }
+        };
+
+        // Now that we know where to put it, let's create a DiscardedEvent
+        let discarded_event = DiscardedEvent {
+            reason: msg.outcome.to_reason().unwrap_or_default().to_string(),
+            category: msg.category,
+            quantity: msg.quantity,
+        };
+        discarded_events.push(discarded_event);
+
+        relay_log::trace!("Sending client report");
+        // Assuming it is not necessary to batch requests here (like HttpOutcomeProducer does),
+        // because outcomes were already aggregated by `OutcomeAggregator`
+        // TODO: verify that client report actually lands in upstream
+        EnvelopeManager::from_registry().do_send(SendClientReport {
+            client_report,
+            scoping: msg.scoping,
+        });
+        Ok(())
+    }
+}
+
+pub struct NonProcessingOutcomeProducer {
+    producer: Option<Recipient<TrackOutcome>>,
+    raw_producer: Option<Recipient<TrackRawOutcome>>,
+}
+
+impl NonProcessingOutcomeProducer {
+    pub fn create(config: Arc<Config>) -> Result<Self, ServerError> {
+        let (producer, raw_producer) = match config.emit_outcomes() {
+            EmitOutcomes::AsOutcomes => {
+                // We emit outcomes as raw outcomes, and accept raw outcomes emitted by downstream
+                // relays.
+                relay_log::info!("Configured to emit outcomes via http");
+                let producer = HttpOutcomeProducer::create(Arc::clone(&config))?.start();
```

Comment:
this is the only time this producer is instantiated, so we can make `HttpOutcomeProducer` private